File: nodejs/second-minimum-node-in-a-binary-tree-671.js
'use strict';
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

let min = ~(1<<31);
let smin = -1;
var findSecondMinimumValue = function (root) {
    if (root===null||root===undefined) {
        return;
    }
    if (root.val < min) {
        smin = min;
        min = root.val;
    }
    else if (root.val < smin && root.val !== min) {
        smin = root.val;
    }
    findSecondMinimumValue(root.left);
    findSecondMinimumValue(root.right);

    return smin === ~(1<<31) ? -1 : smin;

};


function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

console.log("hello world!");

let root = new TreeNode(2);
root.left = new TreeNode(2);
root.right = new TreeNode(5);
// root.left.left = new TreeNode(3);
// root.left.right = new TreeNode(9);
root.right.left = new TreeNode(5);
root.right.right = new TreeNode(7);

console.log(findSecondMinimumValue(root));

File: golang/secondminimumnodeinabinarytree671/second-minimum-node-in-a-binary-tree-671.go
package main

import (
	"fmt"
	"math"
)

func main() {
	root := &TreeNode{Val: 2, Right: &TreeNode{Val: 2}, Left: &TreeNode{Val: 2}}

	fmt.Println(findSecondMinimumValue(root))

	fmt.Println("HELlO world!")
}

var smin = -1
var min = math.MaxInt32

func findSecondMinimumValue(root *TreeNode) int {
	smin = -1
	min = math.MaxInt32

	dfs(root)
	if smin == math.MaxInt32 {
		return -1
	}
	return smin
}

func dfs(root *TreeNode) {
	if root == nil {
		return
	}
	if root.Val < min {
		smin = min
		min = root.Val
	} else if root.Val < smin && root.Val != min {
		smin = root.Val
	}

	dfs(root.Left)
	dfs(root.Right)
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
